README.md

# spaero

[![Travis CI Build Status](https://travis-ci.org/e3bo/spaero.svg?branch=master)](https://travis-ci.org/e3bo/spaero)
[![Coverage Status](https://img.shields.io/codecov/c/github/e3bo/spaero/master.svg)](https://codecov.io/github/e3bo/spaero?branch=master)
[![CRAN Status](https://www.r-pkg.org/badges/version/spaero)](https://cran.r-project.org/package=spaero)

## Overview

`spaero` is being developed to support the work of researchers working
on [Project AERO](https://daphnia.ecology.uga.edu/midas/). `spaero` is
publicly available in part to disseminate and increase the
reproducibility of the project's research and in part because it may
be useful to others.

## Installation

The stable version of `spaero` that is on CRAN may be installed from
within R:

```r
install.packages("spaero")
```

Alternatively, the development version that is on GitHub may be
installed using the `devtools` package:

```r
devtools::install_github("e3bo/spaero")
```

## Documentation

After installation, a listing of the available documentation can be obtained:

```r
help(package = "spaero")
```

## Contributions

Please use GitHub issues or pull requests to let us know about any
bugs you find.

Feature requests are also welcome but be forewarned that we have
limited time to work on items not directly supporting the project's
research.

More details about contribution may be found in the
[contribution guidelines](CONTRIBUTING.md).

## License

GNU GPL version 2 or later


NEWS.md
# spaero 0.5.0

- Add support for pomp version 2.

# spaero 0.4.0

- Change name of "beta" in the parameters of the simulators to
  "beta_par". This change allows the simulation code to be simplified,
  which both resolves an error caused by pomp version 1.19 and may
  prevent future errors as pomp continues to develop.

# spaero 0.3.0

- Add vaccination reaction to simulator. A vaccination rate of zero
  remains the default parameter setting.

- Add backward-looking window option for get_stats.

- Avoid errors when input time series is constant and return a missing
  value instead.

# spaero 0.2.0

- Add transmission argument to create_simulator to allow for
  frequency-dependent transmission. Density-dependent transmission
  remains the default model.

- Add vector of first difference of the variance vector produced by
  get\_stats. This change makes it easier to use the convexity of the
  variance time series as an early warning signal. The name of the
  vector in the stats list is variance\_first\_diff. Note that this
  change makes the abbreviation stats\$var ambiguous. Code using that
  abbreviation to obtain the vector of variance estimates should
  substitute in stats\$variance.

- To the output of get_stats(), add list taus containing Kendall's
  correlation coefficient of the elements of each time series in the
  stats list in the output with time.

- Ensure variance and kurtosis estimates are non-negative. When using
  local linear for estimating statistics, it was possible in previous
  versions for negative values to occur.

# spaero 0.1.1

- Correct autocorrelation calculation. The previous version divided
  the autocovariance by the variance at the most recent time
  point. The current version divides by the geometric mean of the
  variance at each of the two time points, matching standard
  practice. The formula in the vignette for the autocorrelation has
  been changed accordingly.

- Clean up sloppy usage of the term statistic in the documentation.

# spaero 0.1.0

- Initial release


tests/testthat/test.R
if (requireNamespace("lintr", quietly = TRUE)) {
  context("lints")
  test_that("Package Style", {
    lintr::expect_lint_free()
  })
}

set.seed(123)

context("detrending")

test_that("Mean-based detrending works", {
  expect_equal(detrend(1:10, trend="grand_mean")$x, matrix(1:10 - 5.5))
  expect_equal(detrend(cbind(1:10, 2:11), trend="grand_mean")$x,
               cbind(1:10, 2:11) - 6)
  expect_equal(detrend(1:10, trend="ensemble")$x, matrix(rep(0, 10)))
  expect_equal(detrend(cbind(1:10, 2:11), trend="ensemble")$x,
               cbind(1:10, 2:11) - 1:10 - 0.5)
})

test_that("Kernel-based detrending works", {
  x <- runif(1:10)
  x_ind <- seq_along(x)
  expect_equal(
    detrend(x, trend="local_constant", bandwidth=2, kernel="uniform")$x,
matrix(x - stats::ksmooth(x=x_ind, y=x, "box", bandwidth=2, x.points=x_ind)$y))
  expect_equal(
    detrend(x, trend="local_constant", bandwidth=3, kernel="uniform")$x,
matrix(x - stats::ksmooth(x=x_ind, y=x, "box", bandwidth=4, x.points=x_ind)$y))
})

test_that("Skipping detrending works", {
  expect_equal(detrend(1:10, trend="assume_zero")$x, matrix(1:10))
})

if (requireNamespace("np", quietly = TRUE)) {
  if (is.null(options("np.messages")$np.messages)) options(np.messages = TRUE)
  if (is.null(options("np.tree")$np.tree)) options(np.tree = FALSE)

  context("smoothing")

  test_that("Smoothing function works as expected", {
    np_smooth <- function(data, est, bandwidth, kernel="gaussian"){
      is.constant <- grepl("constant", est)
      rt <- ifelse(is.constant, "lc", "ll")
      bw <- np::npregbw(formula=rmn ~ step, bws=bandwidth,
                        regtype=rt, ckertype=kernel,
                        bandwidth.compute=FALSE, data=data,
                        na.action=na.fail)
      mod <- np::npreg(bw)
      list(smooth=fitted(mod), bandwidth=bw$bw)
    }
    data <- data.frame(step=1:10, rmn=1:10)
    expect_equal(smooth(data, est="local_constant", bandwidth=2)$smooth,
                 np_smooth(data, est="local_constant",  bandwidth=2)$smooth)
    expect_equal(smooth(data, est="local_linear", bandwidth=2)$smooth,
                 np_smooth(data, est="local_linear",  bandwidth=2)$smooth,
                 check.names=FALSE)

    expect_equal(smooth(data, est="local_constant", kernel="uniform",
                        bandwidth=3)$smooth,
                 np_smooth(data, est="local_constant", kernel="uniform",
                           bandwidth=3)$smooth)
    expect_equal(smooth(data, est="local_linear", bandwidth=3)$smooth,
                 np_smooth(data, est="local_linear", bandwidth=3)$smooth,
                 check.names=FALSE)

    data2 <- data.frame(step=20:100, rmn=rnorm(81))
    expect_equal(smooth(data2, est="local_constant", bandwidth=5)$smooth,
                 np_smooth(data2, est="local_constant", bandwidth=5)$smooth)
    expect_equal(smooth(data2, est="local_linear", bandwidth=5,
                        kernel="uniform")$smooth,
                 np_smooth(data2, est="local_linear", bandwidth=5,
                           kernel="uniform")$smooth,
                 check.names=FALSE)
    })
}

context("smoothing arguments")

test_that("invalid bandwidths lead to errors", {
  data <- data.frame(step=1:10, rmn=1:10)
  expect_error(smooth(data),
    regexp="argument \"bandwidth\" is missing, with no default")
  expect_error(smooth(data, bandwidth=c(1:20)),
    regexp="argument \"bandwidth\" must be provided as a single numeric value")
  expect_error(smooth(data, bandwidth="hmm"),
    regexp="argument \"bandwidth\" must be provided as a single numeric value")
  expect_error(smooth(data, bandwidth=0.5),
    regexp="argument \"bandwidth\" must be >= 1")
})

context("autocor")

test_that("invalid arguments lead to errors", {
  expect_error(autocor(NA))
  expect_error(autocor(letters), regexp="'x' must be numeric")
  expect_error(autocor(1:10, lag=-1), regexp="'lag' must be >= 0")
})

test_that("large bandwidth autocor estimates agree with acf", {
  w <- rnorm(1000)
  xnext <- function(xlast, w) 0.1 * xlast + w
  x <- Reduce(xnext, x=w, init=0, accumulate=TRUE)

  stats_est <- stats::acf(x, lag.max=1, plot=FALSE)$acf[2, 1, 1]
  spaero_est <- autocor(x, bandwidth=length(x) * 10)$smooth[2]
  expect_equal(stats_est, spaero_est, tolerance=0.05)

  stats_est <- stats::acf(x, lag.max=1, plot=FALSE,
                          type="covariance")$acf[2, 1, 1]
  spaero_est <- autocor(x, bandwidth=length(x) * 10,
                        cortype="covariance")$smooth[2]
  expect_equal(stats_est, spaero_est, tolerance=0.05)

  xnext <- function(xlast, w) 0.9 * xlast + w
  x <- Reduce(xnext, x=w, init=0, accumulate=TRUE)

  stats_est <- stats::acf(x, lag.max=1, plot=FALSE)$acf[2, 1, 1]
  spaero_est <- autocor(x, bandwidth=length(x) * 10)$smooth[2]
  expect_equal(stats_est, spaero_est, tolerance=0.05)
})

context("expected use of get_stats")

test_that(paste("estimate of ensemble stats consistent",
                "in case of time-dependent AR(1) model"), {
  make_time_dependent_updater <- function(f) {
    t <- 0
    updater <- function(xlast, w) {
      phi <- f(t)
      t <<- t + 1
      phi * xlast + w
    }
    return(updater)
  }
  phi_t <- function(t) {
    lambda <- min(-1  + 0.01 * t, 0)
    exp (lambda)
  }
  nensemble <- 3e3
  nobs <- 90
  x <- matrix(nrow=nobs, ncol=nensemble)
  for (i in seq(1, nensemble)){
    updater <- make_time_dependent_updater(f=phi_t)
    w <- rnorm(nobs - 1)
    init <- rnorm(n=1, sd=sqrt(1.16))
    x[, i] <- Reduce(updater, x=w, init=init, accumulate=TRUE)
  }

  est <- get_stats(x, center_trend="assume_zero", stat_bandwidth=3,
                   stat_trend="local_linear")
  lambda_ests <- log(est$stats$autocorrelation[-1])
  lambda_known <- seq(from=-1, by=0.01, len=nobs)
  var_known <-  1 / (1 - exp(2 * lambda_known))
  error_in_mean <- est$stats$mean

  expect_equal(est$centered$x + est$centered$center, x)
  expect_equal(mean(error_in_mean), 0)
  error_in_lambda <- lambda_ests - lambda_known[-1]
  expect_lt(mean(error_in_lambda ^ 2), 0.01)

  ac_error <- est$stats$autocor - exp(lambda_known)
  acov_error <- est$stats$autocov - exp(lambda_known) * var_known
  expect_lt(mean(ac_error ^ 2, na.rm=TRUE), 0.001)
  expect_lt(mean(acov_error ^ 2, na.rm=TRUE), 0.1)

  decay_time_error <- est$stats$decay_time[-1] -  -1 / lambda_known[-1]
  expect_lt((mean(decay_time_error ^ 2)), 0.5)

  expect_lt(mean( (est$stats$var - var_known) ^ 2), 0.1)
  expect_lt(mean(est$stats$skewness ^ 2), 0.01)
  expect_lt(mean( (est$stats$kurtosis - 3) ^ 2), 0.01)

  trend <- sin(2 * pi * (1:nobs) / nobs) + 2
  xx <- x + trend
  est <- get_stats(xx, center_trend="local_constant", center_bandwidth=3,
                   stat_bandwidth=3)
  lambda_ests <- log(est$stats$autocorrelation[-1])

  error_in_mean <- est$stats$mean - trend
  expect_equal(est$centered$x + est$centered$center, xx)
  expect_lt(mean(error_in_mean ^ 2), 0.01)
  error_in_lambda <- lambda_ests - lambda_known[-1]
  expect_lt(mean(error_in_lambda ^ 2), 0.01)

  ac_error <- est$stats$autocor - exp(lambda_known)
  acov_error <- est$stats$autocov - exp(lambda_known) * var_known
  expect_lt(mean(ac_error ^ 2, na.rm=TRUE), 0.001)
  expect_lt(mean(acov_error ^ 2, na.rm=TRUE), 0.1)

  decay_time_error <- est$stats$decay_time[-1] -  -1 / lambda_known[-1]
  expect_lt((mean(decay_time_error ^ 2)), 0.5)

  expect_lt(mean( (est$stats$var - var_known) ^ 2), 0.1)
  error_in_cv <- est$stats$coefficient_of_variation -  sqrt(var_known) / trend
  expect_lt(sqrt(mean( (error_in_cv) ^ 2)), 0.05)
  error_in_id <- est$stats$index_of_dispersion -  var_known / trend
  expect_lt(sqrt(mean( (error_in_id) ^ 2)), 0.2)
  expect_lt(mean(est$stats$skewness ^ 2), 0.01)
  expect_lt(mean( (est$stats$kurtosis - 3) ^ 2), 0.01)
})


test_that(paste("estimate of stats consistent",
                "in case of stationary AR(1) model"), {

  nobs <- 4e3
  w <- rnorm(nobs - 1)
  phi <- 0.1
  xnext <- function(xlast, w) phi * xlast + w
  x <- Reduce(xnext, x=w, init=0, accumulate=TRUE)

  est <- get_stats(x, center_trend="assume_zero", stat_bandwidth=nobs,
                   stat_kernel="uniform", stat_trend="local_linear")
  lambda_ests <- log(est$stats$autocorrelation[-1])
  lambda_known <- rep(log(phi), nobs)
  var_known <-  1 / (1 - exp(2 * lambda_known))
  error_in_mean <- est$stats$mean

  expect_equal(est$centered$x[, 1] + est$centered$center, x)
  expect_equal(mean(error_in_mean), 0)
  error_in_lambda <- lambda_ests - lambda_known[-1]
  expect_lt(mean(error_in_lambda ^ 2), 0.1)

  ac_error <- est$stats$autocor - exp(lambda_known)
  acov_error <- est$stats$autocov - exp(lambda_known) * var_known
  expect_lt(mean(ac_error ^ 2, na.rm=TRUE), 0.002)
  expect_lt(mean(acov_error ^ 2, na.rm=TRUE), 0.1)

  decay_time_error <- est$stats$decay_time[-1] -  -1 / lambda_known[-1]
  expect_lt((mean(decay_time_error ^ 2)), 0.5)

  expect_lt(mean( (est$stats$var - var_known) ^ 2), 0.1)
  expect_lt(mean(est$stats$skewness ^ 2), 0.02)
  expect_lt(mean( (est$stats$kurtosis - 3) ^ 2), 0.05)

  trend <- sin(2 * pi * (1:nobs) / nobs) + 2
  xx <- x + trend
  est <- get_stats(xx, center_trend="local_constant",
                   center_bandwidth=nobs / 16,
                   stat_bandwidth=nobs)
  lambda_ests <- log(est$stats$autocorrelation[-1])

  error_in_mean <- est$stats$mean - trend
  expect_equal(est$centered$x[, 1] + est$centered$center, xx)
  expect_lt(mean(error_in_mean ^ 2), 0.05)
  error_in_lambda <- lambda_ests - lambda_known[-1]
  expect_lt(mean(error_in_lambda ^ 2), 0.2)

  ac_error <- est$stats$autocor - exp(lambda_known)
  acov_error <- est$stats$autocov - exp(lambda_known) * var_known
  expect_lt(mean(ac_error ^ 2, na.rm=TRUE), 0.002)
  expect_lt(mean(acov_error ^ 2, na.rm=TRUE), 0.1)

  decay_time_error <- est$stats$decay_time[-1] -  -1 / lambda_known[-1]
  expect_lt((mean(decay_time_error ^ 2)), 0.5)

  expect_lt(mean( (est$stats$var - var_known) ^ 2), 0.1)
  error_in_cv <- est$stats$coefficient_of_variation -  sqrt(var_known) / trend
  expect_lt(sqrt(mean( (error_in_cv) ^ 2)), 0.2)
  error_in_id <- est$stats$index_of_dispersion -  var_known / trend
  expect_lt(sqrt(mean( (error_in_id) ^ 2)), 0.2)
  expect_lt(mean(est$stats$skewness ^ 2), 0.01)
  expect_lt(mean( (est$stats$kurtosis - 3) ^ 2), 0.02)
})


tests/testthat/test-style.R
context("lints")

test_that("Package Style", {
  skip_if_not_installed("lintr")
  skip_on_cran()
  lintr::expect_lint_free()
})


R/analysis.R
#' Get estimates of time-dependent properties of models.
#'
#' \code{get_stats} estimates time-dependent properties of models
#' (e.g., variance) from ensemble time series.
#'
#' Any missing values in 'x' will cause an error.
#'
#' Bandwidths affect weights in local smoothers as follows. To get the
#' local estimate corresponding to index i, the distance to each other
#' index j is calculated as (i - j) / h, where h is the
#' bandwidth. Then that distance is plugged into the kernel function
#' to obtain a weight. The weights are normalized to sum to one for
#' each index.
#'
#' The gaussian kernel is equivalent to a standard Gaussian density
#' function. The uniform kernel is an indicator function of whether
#' the distance is less than 1. Thus selecting a uniform kernel with a
#' bandwidth of 2 is equivalent to a sliding window of length 3 that
#' is centered on the focal index. In general, if n is the greatest
#' integer that is less than the value of the bandwidth h, the window
#' includes the n nearest values on each side of the focal index.
#'
#' '"local_constant"' smoothers are local means computed with the
#' kernel weights. '"local_linear"' smoothers are the fitted values of
#' local linear regressions with the kernel weights. The linear
#' smoothers avoid biases that the one-sided kernels at the ends of
#' the time series can create for the local constant smoothers.
#'
#' See the vignette "Getting Started with spaero" for the formulas
#' used for each estimate.
#'
#' @param x A univariate or multivariate numeric time series object or
#'     a numeric vector or matrix.
#' @param center_trend Character string giving method of calculating
#'     the trend to subtract. Allowed values are '"assume_zero"',
#'     '"grand_mean"', '"ensemble_means"', '"local_constant"', and
#'     '"local_linear"'. Will be partially matched.
#' @param center_kernel Character string giving the kernel for any
#'     local detrending. Allowed values are '"gaussian"' and
#'     '"uniform"'.
#' @param center_bandwidth Bandwidth of kernel for any local detrending
#'     done. A numeric value >= 1.
#' @param stat_trend Character string giving method of smoothing
#'     estimates. Allowed values are '"local_constant"', and
#'     '"local_linear"'. Will be partially matched.
#' @param stat_kernel Character string giving the kernel for local
#'     smoothing of estimates. Allowed values are '"gaussian"' and
#'     '"uniform"'.
#' @param stat_bandwidth Bandwidth of kernel for local smoothing of
#'     estimates.  A numeric value >= 1.
#' @param lag Integer lag at which to calculate the acf. This lag is
#'     in terms of the index of \code{x} and does not account for the
#'     frequency of \code{x} if \code{x} is a time series. It should
#'     be non-negative.
#' @param backward_only Logical value (defaulting to 'FALSE') that
#'     determines whether any uniform smoothing kernels are restricted
#'     to using data before the index of the smoothed estimate.
#' @param calc_TDAR Logical value (defaulting to 'FALSE') that
#'     determines whether the time-dependent autoregressive model is
#'     fitted.
#' @param tau_bandwidth Size of the window for rolling Kendall's tau
#' calculation.
#'
#' @return A list with elements '"stats"', '"taus"', '"centered"',
#'     '"stat_trend"', '"stat_kernel"', '"stat_bandwidth"', and
#'     '"lag"'. "stats" is a list containing vectors of the
#'     estimates. '"taus"' is a list containing Kendall's correlation
#'     coefficient of each element of '"stats"' with
#'     time. '"centered"' is a list of the detrended time series, the
#'     trend subtracted, and the bandwidth used in the detrending. The
#'     other elements record the parameters provided to this function
#'     for future reference.
#'
#' @seealso \code{\link{acf}}, \code{\link{var}},
#' \code{\link[moments]{kurtosis}}, and
#' \code{\link[moments]{skewness}} for estimation of properties that
#' are not time-dependent. See
#' \code{\link[earlywarnings]{generic_ews}} for another approach to
#' estimation of time-dependent properties.
#' @export
#' @examples
#'
#' # A highly autocorrelated time series
#' x <- 1:10
#' get_stats(x, stat_bandwidth = 3)$stats
#'
#' # Plot log of acf
#' plot(log(get_stats(x, stat_bandwidth = 3)$stats$autocor))
#'
#' # Check estimates with AR1 simulations with lag-1 core 0.1
#' w <- rnorm(1000)
#' xnext <- function(xlast, w) 0.1 * xlast + w
#' x <- Reduce(xnext, x = w, init = 0, accumulate = TRUE)
#' acf(x, lag.max = 1, plot = FALSE)
#' head(get_stats(x, stat_bandwidth = length(x))$stats$autocor)
#'
#' # Check detrending ability
#' x2 <- x + seq(1, 10, len = length(x))
#' ans <- get_stats(x2, center_trend = "local_linear",
#'                   center_bandwidth = length(x),
#'                    stat_bandwidth = length(x))$stats
#' head(ans$autocor)
#'
#' # The simple acf estimate is inflated by the trend
#' acf(x2, lag.max = 1, plot = FALSE)
#'
#'# Check ability to estimate time-dependent autocorrelation
#' xnext <- function(xlast, w) 0.8 * xlast + w
#' xhi <- Reduce(xnext, x = w, init = 0, accumulate = TRUE)
#' acf(xhi, lag.max = 1, plot = FALSE)
#' wt <- seq(0, 1, len = length(x))
#' xdynamic <- wt * xhi + (1 - wt) * x
#' get_stats(xdynamic, stat_bandwidth = 100)$stats$autocor
get_stats <- function(x, center_trend = "grand_mean",
                      center_kernel = c("gaussian", "uniform"),
                      center_bandwidth = NULL,
                      stat_trend = c("local_constant", "local_linear"),
                      stat_kernel = c("uniform", "gaussian"),
                      stat_bandwidth = NULL, lag = 1, backward_only = FALSE,
                      calc_TDAR = FALSE, tau_bandwidth = stat_bandwidth * 5){
  center_kernel <- match.arg(center_kernel)
  stat_kernel <- match.arg(stat_kernel)
  stat_trend <- match.arg(stat_trend)
  centered <- detrend(x, trend = center_trend, kernel = center_kernel,
                      bandwidth = center_bandwidth,
                      backward_only = backward_only)
  stats <- list()


  if (calc_TDAR) {
    f1 <- function(n) return(rep(1, length.out = n))
    f2 <- function(n) return(seq(from = -n / 2 + 0.5, to = n / 2 - 0.5))
    stats$ar_values <- vector()
    for (i in 3:length(x)) { ## TODO What if x is of length 2 or less, or a matrix? Why is lowest i 3 and not 2?
      current_data <- x[1:i]
      new_calculation <- tryCatch({
          M <- get_TDAR_M(y = current_data, f1, f2)
          theta <- get_TDAR_theta(M = M, x = current_data[-1])
          get_TDAR_values(y = current_data, theta = theta, f1, f2)
        }, error = function(e) {
          replace <- 5})
      stats$ar_values[i] <- new_calculation[i]
    }
  }

  stats$variance <- get_noncentral_moments(centered$x, moment_number = 2,
                                           est = stat_trend,
                                           kernel = stat_kernel,
                                           bandwidth = stat_bandwidth,
                                           backward_only = backward_only)
  stats$variance <- stats$variance$smooth
  stats$variance[stats$variance < 0] <- 0
  stats$variance_first_diff <- c(NA, diff(stats$variance))
  stats$autocovariance <- autocor(centered$x, cortype = "covariance", lag = lag,
                                  est = stat_trend, kernel = stat_kernel,
                                  bandwidth = stat_bandwidth,
                                  backward_only = backward_only)
  stats$autocovariance <- stats$autocovariance$smooth
  if (lag > 0) {
    denom <- stats$variance[-seq_len(lag)]
    desel <- seq(length(stats$variance) - lag + 1, length(stats$variance))
    denom <- sqrt(denom * stats$variance[-desel])
    denom <- c(rep(NA, lag), denom)
  } else {
    denom <- stats$variance
  }
  stats$autocorrelation <- stats$autocovariance / denom
  ac01 <- ifelse(0 > stats$autocorrelation, 0, stats$autocorrelation)
  ac01 <- ifelse(1 > ac01, ac01, 1)
  denom <- log(ac01)
  stats$decay_time <- -lag / denom
  stats$mean <- centered$center
  stats$index_of_dispersion <- stats$variance / stats$mean
  stats$coefficient_of_variation <- sqrt(stats$variance) / stats$mean
  stats$skewness <- get_noncentral_moments(centered$x, moment_number = 3,
                                           est = stat_trend,
                                           kernel = stat_kernel,
                                           bandwidth = stat_bandwidth,
                                           backward_only = backward_only)
  stats$skewness <- stats$skewness$smooth / stats$variance ^ (3 / 2)
  stats$kurtosis <- get_noncentral_moments(centered$x, moment_number = 4,
                                           est = stat_trend,
                                           kernel = stat_kernel,
                                           bandwidth = stat_bandwidth,
                                           backward_only = backward_only)
  stats$kurtosis <- stats$kurtosis$smooth / stats$variance ^ 2
  stats$kurtosis[stats$kurtosis < 0] <- 0

  # taus <- lapply(stats, get_tau)
                                        # taus <- lapply(stats, ktseq)
  if (!(backward_only &&  stat_kernel == "uniform")){
    stop("rolling window Kendall's tau is not implemented for these parameters")
  } else {
    window_tau <- function(ewsts) {
      zoo::rollapplyr(ewsts, width = tau_bandwidth, FUN = get_tau, partial = TRUE)
    }
    taus <- lapply(stats, window_tau)
    inputN <- nrow (centered$x)
    windowN <- rep(stat_bandwidth, length.out = inputN)
    windowN[1:stat_bandwidth] <- 1:stat_bandwidth ## partial windows
    pvals <- function(tauts) {
      tryCatch(c(NA, SuppDists::pKendall(q = tauts[-1], N = windowN[-1],
                                         lower.tail = FALSE)),
               error = function(e) NA)
    }
    tau_pvalues <- lapply(taus, pvals)
  }

  ret <- list(stats = stats, taus = taus, centered = centered,
              stat_trend = stat_trend, stat_kernel = stat_kernel,
              stat_bandwidth = stat_bandwidth, lag = lag,
              tau_pvalues = tau_pvalues)
  ret
}

wrap <- function(...){
  unlist(get_stats(...)$taus)
}

#' Get p-values for rolling window correlations of early warning signals
#'
#' \code{get_boot_pvals} estimates the proportion of Kendal's taus
#' that are larger than the estimated values of Kendal's tau for a
#' time series of statistics calculated by \code{get_stats}.
#'
#' Kendal's tau for the null model are simulated by calculating the
#' rolling windows Kendal's tau of rolling window statistics in
#' resamplings of the input time series. A fixed-length--block
#' bootstrap method of resamplings are performed using
#' \code{boot::tsboot}.
#'
#' Rolling window Kendal's tau are only implemented for some choices
#' of parameters for \code{get_stats}, and it is necessary to specify
#' those choices in the \code{...} arguments of this function to avoid
#' a "Not implemented" error. Specifically, one must use
#' '"backward_only" = TRUE'. Also, the flexiblity of
#' \code{get_boot_pvals} is limited and requires that \code{get_stats}
#' calculate a moving window mean as an output statistics as well as
#' the time-dependent autocorrelation. Therefore, the \code{...}
#' arguments should also contain '"calc_TDAR" = TRUE, center_trend =
#' "local_constant"' and provide a value for
#' "center_bandwidth". Examples of suitable calls to
#' \code{get_boot_pvals} are provided below.
#'
#' @param x A univariate numeric time series object or
#'     a numeric vector.
#' @param block_size The block length used in generating replicate
#' time series. 'block_size' should be a postiive integer less tha the
#' number of observations.
#' @param nresamplings The number of bootstrap replicates to use to
#' approximate the p-values.
#' @param parallel Argument for \code{tsboot} to determine type of
#' parallelization used. By default, parallelization is not done.
#'
#' @param ... Named arguments to \code{get_stats} may be supplied here.
#'
#' @return A list with elements names that match the names of '"taus"'
#' in the output of \code{get_stats}. Each element is a vector that
#' gives the p-values of the Kendal's tau correlation coefficients
#' provided by the vectors in the '"taus"' output of \code{get_stats}.
#'
#' @seealso \code{\link{get_stats}} and \code{\link[boot]{tsboot}}
#' @export
#' @examples
#'
#' sdat <- stats::arima.sim(model = list(order = c(2, 0, 0),
#'                          ar = c(0.89, -0.3)), n = 100)
#' bo <- get_boot_pvals(sdat, nresamplings = 100, stat_bandwidth = 10,
#'                      backward_only = TRUE, center_trend = "local_constant",
#'                      center_bandwidth= 10, calc_TDAR = TRUE)
#'
#' foo <- create_simulator(
#'         covar = data.frame(gamma_t = c(0, 0), mu_t = c(0, 0), d_t = c(0, 0),
#'         eta_t = c(0, 0), beta_par_t = c(0, 0.00014), p_t = c(0, 0), time = c(0,
#'           40)))
#' out <- pomp::simulate(foo, times = seq(0, 40, by = 1/26), eta = 1e-3)
#' out <- as(out, "data.frame")
#'
#'
#'
#' head(out)
get_boot_pvals <- function(x, block_size = 52, nresamplings = 300,
                           parallel = "no", ...){
  nms <- names(do.call(get_stats, c(list(x = x), list(...)))$taus)
  N <- length(x)
  bootout <- boot::tsboot(tseries = x, statistic = wrap, R = nresamplings,
                          l = block_size, sim = "fixed", parallel = parallel,
                          ...)
  pvec <- 1 - rowMeans(t(bootout$t) < bootout$t0)
  news <- 11
  stopifnot(length(pvec) == N * news)
  ## assuming there should be a time series of length N for each EWS
  ret <- as.list(as.data.frame(matrix(pvec, nrow = N, ncol = news)))
  names(ret) <- nms
  ret
}

get_tau <- function(x){
  if (any(is.finite(x))){
    stats::cor(x = seq_along(x), y = x, method = "kendall",
               use = "complete.obs")
  } else {
    NA
  }
}

detrend <- function(x, trend = c("grand_mean", "ensemble_means",
                                 "local_constant", "local_linear",
                                 "assume_zero"), bandwidth = NULL, ...){
  trend <- match.arg(trend)
  x <- stats::na.fail(x)
  x <- as.matrix(x)
  if (!is.numeric(x)) stop("'x' must be numeric")
  rmn <- rowMeans(x)
  if (trend == "grand_mean"){
    center <- mean(rmn)
    x <- x - center
  } else if (trend == "ensemble_means"){
    center <- rmn
    x <- x - center
  } else if (trend == "local_constant"
             || trend == "local_linear"){
    samplet <- as.integer(nrow(x))
    step <- seq(1, samplet)
    data <- data.frame(step = step, rmn = rmn)
    srm <- smooth(data = data, bandwidth = bandwidth, est = trend, ...)
    center <- srm$smooth
    x <- x - center
    bandwidth <- srm$bandwidth
  } else if (trend == "assume_zero"){
    center <- rep(0, nrow(x))
  }
  list(x = x, center = center, bandwidth = bandwidth)
}

get_wls_coefs <- function(y, x, w){
  swx <- sum(w * x)
  swy <- sum(w * y)
  sw <- sum(w)
  b <- (sum(w * x * y) - swx * swy / sw) / (sum(w * x ^ 2) - swx ^ 2 / sw )
  a <- (sum(w * y) - b * sum(w * x)) / sw
  c(intercept = a, slope = b)
}

smooth <- function(data, est, kernel = "gaussian", bandwidth,
                   backward_only = FALSE){
  if (!is.numeric(bandwidth) || length(bandwidth) != 1) {
    stop("argument \"bandwidth\" must be provided as a single numeric value")
  } else if (bandwidth < 1){
    stop("argument \"bandwidth\" must be >= 1")
  }
  if (kernel == "gaussian") {
    kern <- function(ind, bw = bandwidth){
      dist <- abs(data$step - ind) / bw
      w <- stats::dnorm(dist)
      w / sum(w)
    }
  } else {
    kern <- function(ind, bw = bandwidth){
      dist <- (data$step - ind) / bw
      if (backward_only) {
        w <- dist <= 0 & dist > -1
      } else {
        w <- dist < 1 & dist > -1
      }
      w / sum(w)
    }
  }
  w <- sapply(data$step, kern)
  if (est == "local_constant"){
    smooth <- colSums(w * data$rmn)
  } else {
    wlm <- function(x){
      coefs <- get_wls_coefs(y = data$rmn, x = data$step, w = w[, x])
      fitted <- data$step * coefs["slope"] + coefs["intercept"]
      fitted[x]
    }
    smooth <- sapply(seq_along(data$step), wlm)
  }
  list(smooth = smooth, bandwidth = bandwidth)
}

autocor <- function(x, cortype = c("correlation", "covariance"), lag = 1, ...){
  cortype <- match.arg(cortype)
  x <- stats::na.fail(x)
  x <- as.matrix(x)
  if (!is.numeric(x)) stop("'x' must be numeric")
  if (lag < 0) stop("'lag' must be >= 0")

  n <- nrow(x)
  end1 <- n - lag
  start2 <- 1 + lag
  x1 <- x[1:end1, , drop = FALSE]
  x2 <- x[start2:n, , drop = FALSE]
  xx_lag <- rowMeans(x1 * x2)

  step <- seq(start2, n)
  data <- data.frame(step = step, rmn = xx_lag)
  xx_lag_sm <- smooth(data = data, ...)
  if (cortype == "correlation") {
    data <- data.frame(step = step, rmn = rowMeans(x1 * x1))
    xx1_sm <- smooth(data = data, ...)
    data <- data.frame(step = step, rmn = rowMeans(x2 * x2))
    xx2_sm <- smooth(data = data, ...)
    denom <- sqrt(xx1_sm$smooth * xx2_sm$smooth)
    ret <- list(smooth = xx_lag_sm$smooth / denom,
                bandwidth = xx1_sm$bandwidth)
  } else {
    ret <- xx_lag_sm
  }
  ret$smooth <- c(rep(NA, lag), ret$smooth)
  ret
}

get_noncentral_moments <- function(x, moment_number = 3, ...) {
  x <- stats::na.fail(x)
  x <- as.matrix(x)
  if (!is.numeric(x)) stop("'x' must be numeric")
  if (moment_number < 1) stop("'moment_number' must be >= 1")

  xpow <- rowMeans(x ^ moment_number)
  step <- seq_along(xpow)
  data <- data.frame(step = step, rmn = xpow)
  smooth(data = data, ...)
}

get_TDAR_M <- function(y, ...) {
  ## use Eq. 12 in BMC Bioinformatics 2008, 9(Suppl 9):S14
  basis_functions <- list(...)

  # Get M
  N <- length(y)
  F <- matrix(nrow = N, ncol = length(basis_functions))
  for (i in 1:length(basis_functions)) {
    F[ , i] <- basis_functions[[i]](N)
  }
  U <- as.numeric(y) * F
  Phi <- U[-N, ]
  x <- y[-1]
  M <- solve(t(Phi) %*% Phi) %*% t(Phi)
  return(M)
}

get_TDAR_theta <- function(M, x) {
  theta <- M %*% x
  return(theta)
}

get_TDAR_values <- function(y, theta, ...) {
  basis_functions <- list(...)
  N <- length(y)
  bf_matrix <- matrix(nrow = N, ncol = length(basis_functions))
  for (i in 1:length(basis_functions)) {
    bf_matrix[ , i] <- basis_functions[[i]](N)*as.numeric(theta[i])
  }
  ar_values <- rowSums(x = bf_matrix)

  return(ar_values)
}
